### ISSUE TITLE
Merging _PortalAuthChallenger into _PortalWrapper

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3295]
|<img alt="tyrion-mx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| tyrion-mx reported|
|-|-|
|Trac ID|trac#3295|
|Type|enhancement|
|Created|2008-06-14 14:30:13Z|
|Branch|https://github.com/twisted/twisted/tree/pb-concurrent-logins-3295|

The login procedure on the server side is actually broken in two step, each in a separate class.
First the client calls the `_PortalWrapper.remote_login` that returns a reference to a new `_PortalAuthChallenger` instance on which the client calls the `remote_respond` method.

They are both needed by the login process and you can't use one of them without the other, so it might be a good idea to merge the `_PortalAuthChallenger` into the `_PortalWrapper` class, as in the attached patch.

Doing this you will have one class instead of two. All the tests in test_pb.py still correctly pass because the `_PortalAuthChallenger` is never used directly by classes other than `_PortalWrapper`.

Attachments:

* [pb.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8b3/8b3ef8e67fb472fcf67a176b26a6722ca3ac372a/029108b5613709738afeeb09fd38e55b104495ef.patch) (1301 bytes) - added by tyrion-mx on 2008-06-14 14:30:59Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3295 3295
type__enhancement enhancement
reporter__tyrion_mx tyrion-mx
priority__lowest lowest
milestone__ 
branch__branches_pb_concurrent_logins_3295 branches/pb-concurrent-logins-3295
branch_author__ 
status__closed closed
resolution__fixed fixed
component__pb pb
keywords__ 
time__1213453813000000 1213453813000000
changetime__1213747455000000 1213747455000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [23928]) Branching to 'pb-concurrent-logins-3295'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [23929]) Add a test for simultaneous pb logins to the same portal wrapper

refs [#3295](https://github.com/twisted/twisted/issues/3295)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

The attached `pb.patch` doesn't break any unit tests, but it does break concurrent pb logins.  Since that means there were no unit tests for concurrent pb logins, I wrote one.  It passes in trunk but fails with the attached patch.  So, I don't think we can merge the two classes, but we should add a test for this functionality.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [23947]) Merge pb-concurrent-logins-3295

Author: exarkun
Reviewer: therve
Fixes: [#3295](https://github.com/twisted/twisted/issues/3295)

Add a unit test for the previously untested feature of
concurrent PB logins to the same root object/portal.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
